logstash-core/lib/logstash/util/wrapped_synchronous_queue.rb

```diff
@@ -218,16 +218,16 @@ def take_originals_from_queue(queue, size, wait)
           if event.nil?
             # queue poll timed out
             next
-          elsif event == LogStash::SHUTDOWN
+          elsif event.is_a?(LogStash::SignalEvent)
             # We MUST break here. If a batch consumes two SHUTDOWN events
             # then another worker may have its SHUTDOWN 'stolen', thus blocking
             # the pipeline.
-            @shutdown_signal_received = true
-            break
-          elsif event == LogStash::FLUSH
+            @shutdown_signal_received = event.shutdown?
```

Comment:
I don't see the purpose of this refactor. the pr description talks about being faster + helps the code in the persistent queue branch.

IMO this could be cleaned as:

```
class SignalEvent;end
class ShutdownEvent < SignalEvent; end
class FlushEvent < SignalEvent; end
```

and then:

```ruby
          case event
          when nil
            # queue poll timed out
            next
          when LogStash::ShutdownEvent
            # We MUST break here. If a batch consumes two SHUTDOWN events
            # then another worker may have its SHUTDOWN 'stolen', thus blocking
            # the pipeline.
            @shutdown_signal_received = true
            break
          when LogStash::FlushEvent
            # See comment above
            # We should stop doing work after flush as well.
            @flush_signal_received = true
            break
          else
            @originals[event] = true
          end
```

Performance wise, in the case of a LogStash::Event the above implementation does .nil? and two === calls, while this PR does .nil?, is_a? and two method calls. The performance should be negligible, and eliminates some lines of code that deal with the flush? and shutdown?.

As for helping the PQ branch..no idea there.